src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -600,12 +867,11 @@ private void handleInsertRowsFailures(
             ERRORS_TOLERANCE_CONFIG,
             ERRORS_DEAD_LETTER_QUEUE_TOPIC_NAME_CONFIG);
       } else {
+        int index = 0;
         for (InsertValidationResponse.InsertError insertError : insertErrors) {
           // Map error row number to index in sinkRecords list.
-          int rowIndexToOriginalSinkRecord = (int) insertError.getRowIndex();
           this.kafkaRecordErrorReporter.reportError(
-              insertedRecordsToBuffer.get(rowIndexToOriginalSinkRecord),
-              insertError.getException());
+              insertedRecordsToBuffer.get(index++), insertError.getException());
```

Comment:
i'm not sure if this is correct, the rowIndex may not be +1 every time?

Reply:
The error and the buffer have been reorganized so that they have a matching index, to wit, record[i] has error[i]

Reply:
It is possible not all rows are erroneous. we should revert back to original code here.